### PR TITLE
[Bugfix] Use 'sum' reduction instead of 'avg' in Async TP reduce-scatter

### PR DESCRIPTION
**Pull request by wangxingran222**
## Purpose

For Row Parallel Linear layers, the partial results from different TP ranks should be summed, not averaged. This PR changes the reduce_op to "sum" to ensure the output magnitude is mathematically correct.

## Test Plan && Test Result

**Test Environment:** H800, TP=4

We tested the accuracy on MMLU and GSM8K. Strangely, on GSM8K, using "avg" outperformed "sum" and the default config (Async TP off). However, we should change "avg" to "sum" to maintain mathematical equivalence.

We also noticed a ~3% accuracy loss on the MMMU-Val benchmark with Qwen3-VL-32B. Note that there are issues with SP + Async TP enabled on multi-modal models in the current vLLM main branch, will address these in a separate PR.

#### SP + Async TP, reduce_op="avg" (Main)

vLLM Serve Command:

```bash
vllm serve Qwen/Qwen3-32B \
        --tensor-parallel-size 4 \
        --port 9000 \
        --no-enable-prefix-caching \
        --compilation-config '{"pass_config": { "enable_sp": true,"fuse_gemm_comms": true}, "splitting_ops":[], "cudagraph_mode": "FULL"}'
```

GSM8K Reuslts:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6566|±  |0.0131|
|     |       |strict-match    |     5|exact_match|↑  |0.7733|±  |0.0115|

MMLU Results:

|      Groups      |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|------------------|------:|------|------|------|---|-----:|---|-----:|
|mmlu              |      2|none  |      |acc   |↑  |0.8052|±  |0.0032|
| - humanities     |      2|none  |      |acc   |↑  |0.7177|±  |0.0063|
| - other          |      2|none  |      |acc   |↑  |0.8326|±  |0.0065|
| - social sciences|      2|none  |      |acc   |↑  |0.8856|±  |0.0056|
| - stem           |      2|none  |      |acc   |↑  |0.8303|±  |0.0065|

#### SP + Async TP, reduce_op="sum" (This PR)

vLLM Serve Command: Same as the "Main" section

GSM8K Reuslts:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6255|±  |0.0133|
|     |       |strict-match    |     5|exact_match|↑  |0.7445|±  |0.0120|

MMLU Results:

|      Groups      |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|------------------|------:|------|------|------|---|-----:|---|-----:|
|mmlu              |      2|none  |      |acc   |↑  |0.8074|±  |0.0032|
| - humanities     |      2|none  |      |acc   |↑  |0.7209|±  |0.0063|
| - other          |      2|none  |      |acc   |↑  |0.8368|±  |0.0064|
| - social sciences|      2|none  |      |acc   |↑  |0.8915|±  |0.0055|
| - stem           |      2|none  |      |acc   |↑  |0.8256|±  |0.0065|

#### TP4

vLLM Serve Command:

```bash
vllm serve Qwen/Qwen3-32B \
        --tensor-parallel-size 4 \
        --port 9000 \
        --no-enable-prefix-caching
```

GSM8K Reuslts:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6240|±  |0.0133|
|     |       |strict-match    |     5|exact_match|↑  |0.7392|±  |0.0121|

MMLU Results:

|      Groups      |Version|Filter|n-shot|Metric|   |Value |   |Stderr|
|------------------|------:|------|------|------|---|-----:|---|-----:|
|mmlu              |      2|none  |      |acc   |↑  |0.8069|±  |0.0032|
| - humanities     |      2|none  |      |acc   |↑  |0.7197|±  |0.0063|
| - other          |      2|none  |      |acc   |↑  |0.8365|±  |0.0064|
| - social sciences|      2|none  |      |acc   |↑  |0.8937|±  |0.0055|
| - stem           |      2|none  |      |acc   |↑  |0.8233|±  |0.0066|

#### SP + Async TP, reduce_op="avg" (Main)

GSM8K Reuslts:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.6247|±  |0.0133|
|     |       |strict-match    |     5|exact_match|↑  |0.7255|±  |0.0123|

#### SP + Async TP, reduce_op="sum" (This PR)

GSM8K Reuslts:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5580|±  |0.0137|
|     |       |strict-match    |     5|exact_match|↑  |0.7005|±  |0.0126|

#### TP4

GSM8K Reuslts:

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.5701|±  |0.0136|
|     |       |strict-match    |     5|exact_match|↑  |0.7005|±  |0.0126|

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

### CODE REVIEW

**Comment by wangxingran222** [comment]
> cc @yifuwang why was the `"avg"` reduction method used in the example?

@ProExpertProg Based on the context from [this PyTorch forum post](https://discuss.pytorch.org/t/distributed-w-torchtitan-introducing-async-tensor-parallelism-in-pytorch/209487#:~:text=fused_matmul_reduce_scatter) regarding TorchTitan, it likely uses 'avg' to sync gradients within the DP group during the backward pass.

I wrote a script to verify the numerical correctness by comparing separate GEMM + Reduce-Scatter against the fused implementation (covering both 'avg' and 'sum' modes).

Specifically, the test setup uses **H800, TP=2**. On each rank, the input tensor is $[4, 3]$ (all ones) and the weight matrix is $[3, 6]$ (all twos). For the **`sum`** reduction, the expected output is a $[2, 6]$ tensor where all values are **12**:

$$
\begin{aligned}
&\text{Given:} \\
&X \in \mathbb{R}^{4 \times 3}, \quad X_{ij} = 1 \\
&W \in \mathbb{R}^{3 \times 6}, \quad W_{ij} = 2 \\
\\
&\text{Local GEMM (on each rank):} \\
&Y = X \times W \quad (\text{Shape: } [4, 6]) \\
&Y_{ij} = \sum_{k=1}^{3} (1 \cdot 2) = 6 \\
\\
&\text{Reduce-Scatter (Sum):} \\
&Z_{\text{global}} = Y_{\text{rank0}} + Y_{\text{rank1}} = 6 + 6 = 12 \\
&Z_{\text{shard}} = \text{Scatter}(Z_{\text{global}}) \quad (\text{Split dim 0}) \\
\\
&\text{Result:} \\
&\text{Shape: } [2, 6], \quad \text{Values: } 12
\end{aligned}
$$

Script:

<details>

```python
import os
import torch
import torch.distributed as dist
import torch.multiprocessing as mp
from torch.distributed._symmetric_memory import enable_symm_mem_for_group

from vllm.distributed import (
    tensor_model_parallel_reduce_scatter,
    get_tp_group
)
from vllm.distributed.parallel_state import (
    init_distributed_environment,
    initialize_model_parallel,
    is_global_first_rank,
)

# -------------------------------------------------------------------------
# Worker Function (Runs per GPU)
# -------------------------------------------------------------------------
def worker_fn(local_rank, world_size, args):
    """
    Simple test comparing fused gemm+rs vs separate gemm+rs
    """
    # 1. Environment Setup
    os.environ["RANK"] = str(local_rank)
    os.environ["LOCAL_RANK"] = str(local_rank)
    os.environ["WORLD_SIZE"] = str(world_size)
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "29500"

    device = torch.device(f"cuda:{local_rank}")
    torch.cuda.set_device(device)

    # 2. Init vLLM Distributed
    init_distributed_environment()
    initialize_model_parallel(tensor_model_parallel_size=world_size)

    # Enable symmetric memory for the TP process group
    enable_symm_mem_for_group(get_tp_group().device_group.group_name)

    # 3. Create simple test data
    # Input tensor: all ones (shape: 4x3)
    input_tensor = torch.ones((4, 3), dtype=torch.float32, device=device)
    
    # Weight matrix: all twos (shape: 3x6)
    weight_matrix = torch.ones((3, 6), dtype=torch.float32, device=device) * 2
    
    # 4. Method 1: Separate GEMM + Reduce Scatter
    # GEMM
    gemm_result = torch.mm(input_tensor, weight_matrix)
    
    # Reduce Scatter
    rs_result = tensor_model_parallel_reduce_scatter(gemm_result, dim=0)
    
    # 5. Method 2: Fused GEMM + Reduce Scatter (avg)
    tp = get_tp_group()
    
    fused_result_avg = torch.ops.symm_mem.fused_matmul_reduce_scatter(
        input_tensor,
        weight_matrix,
        "avg",
        scatter_dim=0,
        group_name=tp.device_group.group_name,
    )
    
    # 6. Method 3: Fused GEMM + Reduce Scatter (sum)
    fused_result_sum = torch.ops.symm_mem.fused_matmul_reduce_scatter(
        input_tensor,
        weight_matrix,
        "sum",
        scatter_dim=0,
        group_name=tp.device_group.group_name,
    )
    
    # 7. Compare results
    dist.barrier()
    
    if is_global_first_rank():
        print("\n" + "="*60)
        print(" [COMPARISON] Separate vs Fused GEMM+RS")
        print("="*60)
        print(f"Input tensor shape: {input_tensor.shape}")
        print(f"Weight matrix shape: {weight_matrix.shape}")
        print(f"GEMM result shape: {gemm_result.shape}")
        print(f"Reduce Scatter result shape: {rs_result.shape}")
        print(f"Fused GEMM+RS (avg) result shape: {fused_result_avg.shape}")
        print(f"Fused GEMM+RS (sum) result shape: {fused_result_sum.shape}")
        print("-" * 60)
        print(f"Baseline (Separate GEMM+RS) result:\n{rs_result}")
        print(f"Fused GEMM+RS (avg) result:\n{fused_result_avg}")
        print(f"Fused GEMM+RS (sum) result:\n{fused_result_sum}")
        print("-" * 60)
        
        # Compare baseline vs fused (avg)
        diff_avg = (rs_result - fused_result_avg).abs()
        max_diff_avg = diff_avg.max().item()
        mean_diff_avg = diff_avg.mean().item()
        
        print("Baseline vs Fused (avg):")
        print(f"  Max difference: {max_diff_avg:.6e}")
        print(f"  Mean difference: {mean_diff_avg:.6e}")
        if max_diff_avg < 1e-5:
            print("  ✓ Results match!")
        else:
            print("  ✗ Results differ!")
        
        print()
        
        # Compare baseline vs fused (sum)
        diff_sum = (rs_result - fused_result_sum).abs()
        max_diff_sum = diff_sum.max().item()
        mean_diff_sum = diff_sum.mean().item()
        
        print("Baseline vs Fused (sum):")
        print(f"  Max difference: {max_diff_sum:.6e}")
        print(f"  Mean difference: {mean_diff_sum:.6e}")
        if max_diff_sum < 1e-5:
            print("  ✓ Results match!")
        else:
            print("  ✗ Results differ!")
        
        print("="*60 + "\n")
    
    # 8. Cleanup resources
    torch.cuda.synchronize()
    torch.cuda.empty_cache()
    dist.destroy_process_group()

if __name__ == "__main__":
    WORLD_SIZE = 2

    print(f"Starting Simple GEMM+RS Comparison Test on {WORLD_SIZE} GPUs...")
    
    mp.spawn(
        worker_fn,
        args=(WORLD_SIZE, {}),
        nprocs=WORLD_SIZE,
        join=True
    )
```

</details>

Results:

<details>

```
============================================================
 [COMPARISON] Separate vs Fused GEMM+RS
============================================================
Input tensor shape: torch.Size([4, 3])
Weight matrix shape: torch.Size([3, 6])
GEMM result shape: torch.Size([4, 6])
Reduce Scatter result shape: torch.Size([2, 6])
Fused GEMM+RS (avg) result shape: torch.Size([2, 6])
Fused GEMM+RS (sum) result shape: torch.Size([2, 6])
------------------------------------------------------------
Baseline (Separate GEMM+RS) result:
tensor([[12., 12., 12., 12., 12., 12.],
        [12., 12., 12., 12., 12., 12.]], device='cuda:0')
Fused GEMM+RS (avg) result:
tensor([[6., 6., 6., 6., 6., 6.],
        [6., 6., 6., 6., 6., 6.]], device='cuda:0')
Fused GEMM+RS (sum) result:
tensor([[12., 12., 12., 12., 12., 12.],
        [12., 12., 12., 12., 12., 12.]], device='cuda:0')
------------------------------------------------------------
Baseline vs Fused (avg):
  Max difference: 6.000000e+00
  Mean difference: 6.000000e+00
  ✗ Results differ!

Baseline vs Fused (sum):
  Max difference: 0.000000e+00
  Mean difference: 0.000000e+00
  ✓ Results match!
============================================================
```

</details>

Additionally, we can reference PyTorch's native [Async TP Compile Pass](https://github.com/pytorch/pytorch/blob/d924a23ce2395e6b4f82111bfc8d452e2550013a/torch/_inductor/fx_passes/micro_pipeline_tp.py#L862
). It does **not** hardcode `'avg'` as the reduction operation. Instead, it dynamically detects and preserves the `reduce_op` used in the original reduce-scatter node.

In vLLM, calls to `allreduce` and `reduce-scatter` generally do not explicitly pass a `reduce_op` argument. Therefore, they rely on the default behavior defined in the [signature](https://github.com/pytorch/pytorch/blob/d924a23ce2395e6b4f82111bfc8d452e2550013a/torch/distributed/distributed_c10d.py#L4504), which is `ReduceOp.SUM`.

```python
def reduce_scatter_tensor(output, input, op=ReduceOp.SUM, group=None, async_op=False):
```

**Review by ProExpertProg — APPROVED** [review]
SOunds good, please resolve the merge conflicts then we can merge!
